runtime/ints.c

```diff
@@ -280,6 +280,157 @@ CAMLprim value caml_int32_shift_right(value v1, value v2)
 CAMLprim value caml_int32_shift_right_unsigned(value v1, value v2)
 { return caml_copy_int32((uint32_t)Int32_val(v1) >> Int_val(v2)); }
 
+#if defined(__GNUC__)
+#if ARCH_INT32_TYPE == long
+#define int32_clz __builtin_clzl
+#define int32_popcnt __builtin_popcountl
+#else /* ARCH_INT32_TYPE == long */
+#define int32_clz __builtin_clz
+#define int32_popcnt __builtin_popcount
+#endif /* ARCH_INT32_TYPE == long */
+
+#define int64_clz __builtin_clzll
+#define int64_popcnt __builtin_popcountll
+
+#else /* defined(__GNUC__) */
+ /* _MSC_ does not have builtin for popcnt, and
+  *  has undefined behavior for clz builtin when
+  *  lzcnt instruction is not supported.
+  *  Use naive version of clz and popcnt from Hacker's Delight. */
+
+int naive_int64_clz(uint64_t v)
+{
+  int n = 0;
+  int64_t x = (int64_t) v;
+  int64_t y = x;
+
+L: if (x < 0) return n;
+   if (y == 0) return 64 - n;
+   n = n + 1;
+   x = x << 1;
```

Comment:
This has undefined behaviour for large values of `x`.

Reply:
I'll change it. 

I am not sure it is undefined in this particular case, because of the check x < 0 on line 307 guarantees that x is nonnegative on line 310.  From C99, it appears to be fine, but very subtle and better not to get into this corner case. Thank you for pointing it out!

 "If E1 has a signed type and nonnegative value, and E1 × 2 E2 is representable in the result type, then that is the resulting value; otherwise, the behavior is undefined."

Reply:
ah, I see - the very last shift is undefined.